_posts/2022-05-17-data-location-inheritance-bug.md

```diff
@@ -0,0 +1,64 @@
+---
+layout: post
+published: true
+title: 'Bug Concerning Data Location during Inheritance'
+date: '2022-05-17'
+author: Solidity Team
+category: Security Alerts
+---
+
+On February 5th 2021, Nicolas Venturo reported a bug that allows
+overriding functions to change the data location of parameters from
+``memory`` to ``calldata``.
+
+TODO explain why it took us over a year to fix it.
+
+The bug was introduced in Solidity 0.6.9 together with the ability to use ``calldata``
+data location for all variables (and not just parameters of external functions).
+
+We assigned the bug a severity of "low".
+
+## Which Contracts are Affected?
+
+The effect of the bug is that a memory pointer is interpreted as a calldata pointer
+or vice-versa. It can only happen if you change the data location of a function
+during inheritance and perform a call at a location that is only aware of
+the function in the base contract (i.e. inside the base contract itself
+or a contract inheriting only from it, but not from the derived).
+
+```solidity
+abstract contract I {
+    // The base contract uses "calldata"
+    function f(uint[] calldata x) virtual internal;
+}
+contract C is I {
+    // The derived contract uses "memory" and the compiler
+    // does not complain - this is the bug in the compiler.
+    function f(uint[] memory x)  override internal {
+        // If you use ``x``, it will access memory
+        // even if "D.g" passed us a calldata pointer.
```

Comment:
For consistency with the comment below:
```suggestion
        // If you use `x`, it will access memory
        // even if `D.g` passed us a calldata pointer.
```